### PR TITLE
3841 `ZStream#fromIterator(Managed)` should have an optional `chunkSize`

### PR DESCRIPTION
**Pull request by ogresto**
…arameter #3841

### CODE REVIEW

**Comment by ogresto** [comment]
https://github.com/zio/zio/issues/3841

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -3424,33 +3424,39 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
...
-        ZStream.repeatEffectOption {
-          Task {
-            val hasNext: Boolean =
-              try it.hasNext
-              catch {
-                case e: Throwable if !rt.platform.fatal(e) =>
-                  throw e
-              }
+    ZStream
+      .fromEffect(Task(iterator.grouped(chunkSize)) <*> ZIO.runtime[Any])
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:3434]: Great use of `grouped` 👍

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -3424,33 +3424,39 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
...
+      .flatMap {
+        case (it, rt) =>
+          ZStream.repeatEffectOption {
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:3437]: Let's use `repeatEffectChunkOption` - should be slightly more efficient

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 3]
```diff
@@ -3424,33 +3424,39 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
...
-              catch {
-                case e: Throwable if !rt.platform.fatal(e) =>
-                  throw e
-              }
-            } else throw StreamEnd
-          }.mapError {
-            case StreamEnd => None
-            case e         => Some(e)
+              if (hasNext) {
+                try it.next()
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:3447]: You can do `Chunk.fromIterable(it.next())` here

**Review by iravid — COMMENTED** [review]
Looks great! Some minor efficiency comments.

**Comment by ogresto** [comment]
@iravid Why do we need an optional `chunkSize` parameter if ZStream offers inherent laziness (as the documentation says) ?

**Comment by iravid** [comment]
> @iravid Why do we need an optional `chunkSize` parameter if Iterator is already lazy?

If I understand the question, Chunks are used in ZStream for performance reasons. Most pure operations benefit from being performed on chunk in a tight loop.

**Comment by ogresto** [comment]
'This job has been blocked because resource-class:large is not available on your plan. Please upgrade to continue building.' @iravid @simpadjo could you please upgrade ci plan ?

**Comment by ogresto** [comment]
> > @iravid Why do we need an optional `chunkSize` parameter if Iterator is already lazy?
> 
> If I understand the question, Chunks are used in ZStream for performance reasons. Most pure operations benefit from being performed on chunk in a tight loop.

i am not sure this is a problem, but 'grouped' iterator is not lazy anymore (under the hood it initiate values batch-by-batch as an array). And it breaks some tests (for example 
`testM("toIterator") {
          (for {
            counter  <- Ref.make(0).toManaged_ //Increment and get the value
            effect    = counter.updateAndGet(_ + 1)
            iterator <- ZStream.repeatEffect(effect).toIterator
            n         = 2000
            out <- ZStream
                     .fromIterator(iterator.map(_.merge))
                     .mapConcatM(element => effect.map(newElement => List(element, newElement)))
                     .take(n.toLong)
                     .runCollect
                     .toManaged_
          } yield assert(out)(equalTo(Chunk.fromIterable(1 to n)))).use(ZIO.succeed(_))
        }`
)

**Comment by adamgfraser** [comment]
@ogresto That warning is being generated because it looks like the CI is being run on your Circle CI account instead of the one for the ZIO organization. We've seen this a couple times before. @softinio Do you remember how we resolved this?

**Comment by iravid** [comment]
@ogresto Could you try the following:
- Close this PR and open a new one
- Rename your fork of ZIO from `fzio` to `zio`

**Comment by ogresto** [comment]
@iravid it seems it doesn't work

**Pull request closed without merging**